rfc/008-platform-support.md

```diff
@@ -0,0 +1,111 @@
+current implementation doesn't distinguish between platform features
+(for example which hypervisor is used) but only between full platform combinations.
+
+[node-installer]: https://github.com/edgelesssys/contrast/blob/3f39682ea9a383b2557923e257bd065e461b8ee6/nodeinstaller/internal/constants/constants.go#L48
+
+## Proposal
+
+### Structured Representation
+
+Throughout the implementation code, platforms shouldn't be represented as _individual combinations_,
+but rather through a generic, structured type that contains information about their _features_.
+
+So `AKS-CLH-SNP` could become something like this:
+
+```go
+type Platform struct {
+    KubernetesDistribution KubernetesDistribution
+    Hypervisor Hypervisor
+    HardwarePlatform HardwarePlatform
+}
+
+// AKS-CLH-SNP
+foo := Platform{
+    KubernetesDistribution: AKS,
+    Hypervisor: CloudHypervisor,
+    HardwarePlatform: SNPMilan,
+}
+```
+
+This would allow the implementation code (for example the aforementioned switch for [Kata configuration])
+to be reduced to only depend on the information bits it requires.
+
+For example, this:
+
+```go
+case platforms.MetalQEMUTDX, platforms.K3sQEMUTDX, platforms.RKE2QEMUTDX:
+```
+
+Could become this:
+
+```go
+case TDX
+```
+
+This would also enable easy validation of user error, as the individual types can be checked for
+validity at parsing time, making invalid states unrepresentable [^1].
+
+However, to get to such a structured format, the information needs to be passed to Contrast in another
+structured format, such as a configuration file or another canonical serialized format, which should
+be discussed in the following section.
+
+[Kata configuration]: https://github.com/edgelesssys/contrast/blob/3f39682ea9a383b2557923e257bd065e461b8ee6/nodeinstaller/internal/constants/constants.go#L48
+
+### Serialization
+
+As platforms are passed throughout multiple (programming) languages, such as the `justfile` and Go code for
+developers, but - in the most primitive way - from a user invoking the CLI in a shell to Go code, there must
+be a serializable, machine-readable representation of a platform.
+
+This could be a configuration file that's read by the CLI and other dependants.
+
+This would reduce implementation complexity, as Contrast could rely on existing YAML / TOML / JSON / etc. parsing
+libraries and no canonical string representation of a platform would need to be implemented.
+
+Additionally, adding platform features would not lead to unreasonably large string representations, which might
+be hard to read for humans at some point.
+
+## Considerations
+
+### Node-Installer Images
+
+As it would be complex and costly to publish node-installer images for all supported platform
+combinations, it might be better to ship a singular one-size-fits-all node installer with support
+for all platforms. This could also be a foundation for heterogeneous deployments that utilize multiple
```

Comment:
Suggestion on how to solve this: Remove all the stuff from the node installer image, give it a list of url/hash pairs. Send the config to the node-installer (via cm or what) and let it download the required components one by one.

Reply:
This makes installation in an air-gapped / proxied facility less convenient.

Reply:
we can fetch from a file path as fallback and ship a separate image for air gapped, then we cover both with two images.

Reply:
I thought about this, but we also have to consider that it brings quite some overhead to development, as we lose the benefit of a self-contained image, and have to build some kind of uploading pipeline again for the components to S3 or so.

Reply:
we could also push/pull the artifacts from a container registry with oras.
I think the dev workflow will be more complex the more variants we have either way, main goal should be to make this convenient for the user while keeping it manageable for devs.

Reply:
I think both an agnostic monolithic node-installer, as well as a component-based node-installer have their up- and downsides for devs _and_ users. We'll need to think through the individual possibilities here together.